### PR TITLE
Add Sphinx-gallery to doc

### PR DESCRIPTION
**Pull request by mthrok**
Add Sphinx-gallery to documentation so that we can maintain the notebook for the main branch, and copy it over to tutorial repo at new releases.

https://419306-90321822-gh.circle-artifacts.com/0/docs/index.html

## Basic Goal

- [x] Add basic sphinx-gallery
  - Note: so as to make it easy to copy the gallery files to tutorial repo at release, we mimic their config as much as possible. 
- [x] Port forced alignment example (and fix some minor issues)
  - [x] Fix the last plot which SG does not render properly
- [x] Port ASR example (and fix)

## Stretch Goal (separate PRs)

- [x] Add `Run in Google Colab` [[src](https://github.com/pytorch/pytorch_sphinx_theme/blob/fe1f3d5b9233497d81d04f55f5750ccad92500be/pytorch_sphinx_theme/layout.html#L275-L297)] (might be better to update PyTorch-sphinx-theme) (#1969)
   - [x] Fix the link prefix (`pytorch/tutorials`)
- [ ] Port Audio manipulation examples #1970 
- [ ] Port TTS examples  #1973 
- [ ] Port Speech command classification example
- [ ] Port MVDR example

### CODE REVIEW

File: docs/source/functional.rst [hunk 1]
```diff
@@ -4,6 +4,8 @@
 torchaudio.functional
 =====================
 
+.. py:module:: torchaudio.functional
+
```
Comment (nateanl) [inline docs/source/functional.rst:8]: What's the function of adding it for sphinx-gallery?

File: examples/gallery/wav2vec2/speech_recognition_pipeline_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+# prior probability distribution are differnt (in typical conversations,
+# ``night`` would occur way more often than ``knight``), to accurately
+# generate transcripts with ``knight``, such as ``a knight with a sword``,
+# the decoding process has to postpone the final decision until it sees
+# enough context.
+# 
+# There are many decoding techniques proposed, and they require external
+# resources, such as word dictionary and language models.
+# 
+# In this tutorial, for the sake of simplicity, we will perform greeding
```
Comment (nateanl) [inline examples/gallery/wav2vec2/speech_recognition_pipeline_tutorial.py:234]:
```suggestion
# In this tutorial, for the sake of simplicity, we will perform greedy
```

File: examples/gallery/wav2vec2/speech_recognition_pipeline_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+class GreedyCTCDecoder(torch.nn.Module):
+  def __init__(self, labels, blank=0):
+    super().__init__()
+    self.labels = labels
+    self.blank = blank
+
+  def forward(self, emission: torch.Tensor) -> str:
+    """Given a sequence emission over labels, get the best path string
+    Args:
+      emission (Tensor): Logit tensors. Shape `[num_seq, num_label]`.
```
Comment (nateanl) [inline examples/gallery/wav2vec2/speech_recognition_pipeline_tutorial.py:251]:
```suggestion
      emission (Tensor): Logit tensors. Shape `[num_seq, num_label]`.

```

File: examples/gallery/wav2vec2/speech_recognition_pipeline_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+######################################################################
+# Now create the decoder object and decode the transcript.
+# 
+
+decoder = GreedyCTCDecoder(labels=bundle.get_labels())
+transcript = decoder(emission[0])
+
+
+######################################################################
+# Let’s check the result and listen again the audio.
```
Comment (nateanl) [inline examples/gallery/wav2vec2/speech_recognition_pipeline_tutorial.py:270]:
```suggestion
# Let’s check the result and listen again to the audio.
```

File: examples/gallery/wav2vec2/speech_recognition_pipeline_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+# 
+
+print(transcript)
+IPython.display.display(IPython.display.Audio(SPEECH_FILE))
+
+
+######################################################################
+# The ASR model is fine-tuned using a loss function called CTC.
```
Comment (nateanl) [inline examples/gallery/wav2vec2/speech_recognition_pipeline_tutorial.py:278]: maybe write the full name 
```suggestion
# The ASR model is fine-tuned using a loss function called Connectionist Temporal Classification (CTC).
```

**Comment by nateanl** [comment]
The tutorial has the same rendering issue for the last plot: https://pytorch.org/tutorials/intermediate/forced_alignment_with_torchaudio_tutorial.html
We can create a PR there and fix it.

File: docs/source/functional.rst (hunk 1 above)
Comment (mthrok) [reply docs/source/functional.rst:8]: This allows to refer to the module as a while from examples, like `:py:mod:`torchaudio.functional`.

**Comment by mthrok** [comment]
> The tutorial has the same rendering issue for the last plot: https://pytorch.org/tutorials/intermediate/forced_alignment_with_torchaudio_tutorial.html We can create a PR there and fix it.

@nateanl Yeah, once this one is merged, I will make a PR in tutorials repo to port the fix.

File: docs/source/functional.rst (hunk 1 above)
Comment (nateanl) [reply docs/source/functional.rst:8]: Got it. Looks nice.

**Review by nateanl — APPROVED** [review]

**Comment by carolineechen** [comment]
Overall the changes/rendering looks good, but I was originally under the assumption that the gallery was only for tutorials that aren't also in the pytorch.org tutorials section (either because they use unreleased features or if for some reason we do not think it belongs in pytorch.org). What is the reason for including the same tutorials in both the gallery and website? Updating existing tutorials would then require twice the amount of work to ensure that they are synced.

On another note, I realized the W2V2 speech recognition tutorial link for pytorch.org seems broken, at least on my end https://pytorch.org/tutorials/intermediate_source/speech_recognition_pipeline_tutorial.html

**Comment by mthrok** [comment]
> Overall the changes/rendering looks good, but I was originally under the assumption that the gallery was only for tutorials that aren't also in the pytorch.org tutorials section (either because they use unreleased features or if for some reason we do not think it belongs in pytorch.org). What is the reason for including the same tutorials in both the gallery and website? Updating existing tutorials would then require twice the amount of work to ensure that they are synced.
> 
> On another note, I realized the W2V2 speech recognition tutorial link for pytorch.org seems broken, at least on my end https://pytorch.org/tutorials/intermediate_source/speech_recognition_pipeline_tutorial.html

There are couple of advantages of maintaining the tutorial here as well.

1. We can keep the corresponding tutorials up-to-date with main branch.
    With the recent BC-breaking change on wav2vec2 pipeline (#1914), we will need to update the speech recognition tutorial. By having the tutorial in our repo, we can update the tutorial as we land the change. If we only have the tutorials, then we need to remember to update the tutorial until right before the release, when they prepare the release branc on repo, which is very inconvenient. (This is why I put the note https://github.com/pytorch/audio/issues/1950) 
2. Updating the tutorials at release will be just copy-ing the corresponding files to tutorial repo.
    I am setting up the gallery mimicking the configuration of tutorial repo, so that updating the tutorials in tutorial repo is just about replacing it with the ones kept in audio repo. Sure this adds a slight overhead, but I do not think it doubles the efforts.
3. The turn-around time is much better
    With our repo only containing the tutorials of our own, the time it takes for building/debugging the tutorial (the combination of Python code and rST formats, which is not what we are used to write) is much shorter. And it is much easier to debug the issue with the tutorial, (such as the rendering issue in forced alignment tutorial that I did not notice on tutorial repo because turn-around time was too long, but fixed in this PR kinda quickly) 

So in short, we can consider the tutorials in our repo as a staging place with nightly binary packages.

**Comment by mthrok** [comment]
> On another note, I realized the W2V2 speech recognition tutorial link for pytorch.org seems broken, at least on my end https://pytorch.org/tutorials/intermediate_source/speech_recognition_pipeline_tutorial.html

Where did you get this?

The tutorial is at 
https://pytorch.org/tutorials/intermediate/speech_recognition_pipeline_tutorial.html

**Comment by nateanl** [comment]
https://pytorch.org/tutorials/ and click the asr demo. I got the same 404 as @carolineechen.

**Comment by carolineechen** [comment]
@mthrok ok, that makes sense to me. In that case we just need to make sure to remember to add to the release milestone or create a tutorials label to keep track of the necessary tutorial updates prior to release.

and yea, the tutorial link you commented works, but clicking through pytorch.org/tutorials -> audio -> ASR tutorial redirects to the broken link I commented

**Comment by mthrok** [comment]
@carolineechen  @nateanl  Okay, the fix for the tutorial link https://github.com/pytorch/tutorials/pull/1738

**Review by carolineechen — APPROVED** [review]

**Comment by mthrok** [comment]
Thanks!
